### PR TITLE
ENH: Pythran implementation of _cdf_distance

### PR DESCRIPTION
**Pull request by charlotte12l**
#### What does this implement/fix?
Pythran implementation of _cdf_distance.

```
u_values, v_values = [0.7, 7.4, 2.4, 6.8], [1.4, 8. ] 
u_weights, v_weights = [2.1, 4.2, 7.4, 8. ], [7.6, 8.8]

u_values = np.asarray(u_values, dtype=float)
u_weights = np.asarray(u_weights, dtype=float)

v_values = np.asarray(v_values, dtype=float)
v_weights = np.asarray(v_weights, dtype=float)
```

The Pythran version:
```
%timeit -n 100 -r 3 _pythran_cdf_distance(1,u_values, v_values, u_weights, v_weights)
23.4 µs ± 8.47 µs per loop (mean ± std. dev. of 3 runs, 100 loops each)
```
The original version:
```
%timeit -n 100 -r 3 _orig_cdf_distance(1,u_values, v_values, u_weights, v_weights )
74.4 µs ± 23.9 µs per loop (mean ± std. dev. of 3 runs, 100 loops each)
```

3x speedup with a small array input.

#### I also rewrote some parts to support Pythran compilation

Rewrote `u_values[u_sorter].searchsort` to `np.searchsort` to avoid Pythran `AssertionError: Function path is chained attributes and name.`

Rewrote `all_values.sort(kind='mergesort')` to `np.sort(all_values, kind='mergesort')` to avoid Pythran compliation error.

cc @rgommers @serge-sans-paille

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
Great steps @charlotte12l !

Some minor comments:

> Rewrote u_values[u_sorter].searchsort to np.searchsort to avoid Pythran AssertionError: Function path is chained attributes and name.
Can you create a minimal reproducer and sublit an issue on the Pythran tracker? Such feedbacks are important to us :-)

> Rewrote all_values.sort(kind='mergesort') to np.sort(all_values, kind='mergesort') to avoid Pythran compliation error.
Same :-)

File: scipy/stats/_cdf_distance_1d.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import numpy as np
+
+#pythran export _cdf_distance_1d(int, float[:], float[:])
+#pythran export _cdf_distance_1d(int, float[:], float[:], None, None)
+#pythran export _cdf_distance_1d(int, float[:], float[:], float[:], float[:])
```
Comment (serge-sans-paille) [inline scipy/stats/_cdf_distance_1d.py:5]: this should also work (and catch more cases):
```
#pythran export _cdf_distance_1d(int, float[:], float[:], float[:] or None, float[:] or None)
```

**Comment by charlotte12l** [comment]
@serge-sans-paille 
> Can you create a minimal reproducer and sublit an issue on the Pythran tracker? Such feedbacks are important to us :-)

Sure! I just opened two issues on Pythran: https://github.com/serge-sans-paille/pythran/issues/1791, https://github.com/serge-sans-paille/pythran/issues/1792

File: scipy/stats/_cdf_distance_1d.py (hunk 1 above)
Comment (charlotte12l) [reply scipy/stats/_cdf_distance_1d.py:7]: Oh I forgot these two cases. Thank you very much !!

File: scipy/stats/_cdf_distance_1d.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    all_values = np.concatenate((u_values, v_values))
+    all_values = np.sort(all_values, kind='mergesort')
+
+    # Compute the differences between pairs of successive values of u and v.
+    deltas = np.diff(all_values)
+
+    # Get the respective positions of the values of u and v among the values of
+    # both distributions.
+    u_cdf_indices = np.searchsorted(u_values[u_sorter],all_values[:-1],'right')
+    v_cdf_indices = np.searchsorted(v_values[v_sorter],all_values[:-1],'right')
```
Comment (rgommers) [inline scipy/stats/_cdf_distance_1d.py:65]: Minor: if you're touching this code anyway, can you add some PEP 8 spaces after the commas?

**Review by rgommers — COMMENTED** [review]
This looks good, thanks @charlotte12l!

The two public functions that currently make use of `_cdf_distance` are `wasserstein_distance` and `energy_distance`. Those don't have benchmarks yet, so it would be good to add some to show that these functions also get significant speedups.

> Rewrote `u_values[u_sorter].searchsort` to np.searchsort` to avoid Pythran `AssertionError: Function path is chained attributes and name`.

This is completely equivalent, and I'd say using the function is generally preferred over the method anyway.

> Rewrote `all_values.sort(kind='mergesort')` to `np.sort(all_values, kind='mergesort')` to avoid Pythran compliation error.

This is a change from in-place to out-of-place sorting - but the code is equivalent, and the new version is faster. So that's all good as well.

**Comment by rgommers** [comment]
CI failures are all unrelated.

File: scipy/stats/_cdf_distance_1d.py (hunk 2 above)
Comment (charlotte12l) [reply scipy/stats/_cdf_distance_1d.py]: Sorry, I forgot that before. Just added the spaces.

**Comment by charlotte12l** [comment]
@rgommers 

> The two public functions that currently make use of `_cdf_distance` are `wasserstein_distance` and `energy_distance`. Those don't have benchmarks yet, so it would be good to add some to show that these functions also get significant speedups.

Just added the benchmarks: https://github.com/scipy/scipy/pull/14163 :)

**Comment by mdhaber** [comment]
Given the merge conflicts, the fact that benchmarks are needed, the inactivity, and the difficulty we're having in gh-13957, I'm going to go ahead and close this one. Thanks all - and feel free to reopen if you can help finish it up.

**Pull request closed without merging**